### PR TITLE
grpc_json_transcoder: Adhere to decoder and encoder buffer limits

### PR DESCRIPTION
**Pull request by nareddyt**
Commit Message:
grpc_json_transcoder: Adhere to decoder and encoder buffer limits

Additional Description:
Currently, the transcoder has unbounded internal buffering. This is not allowed according to the [Extension stability and security posture](https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#extension-stability-and-security-posture).

This PR introduces limits to internal request/response buffers. This follows the documentation for [Flow Control](https://github.com/envoyproxy/envoy/blob/main/source/docs/flow_control.md), but the filter does not participate in watermarking. Flow control via watermarking will not help the filter: Data stuck in a buffer will never be transcoded until more data comes in. If watermarking occurs, then the data will be stuck forever. So we fail the request with `HTTP 413` or `HTTP 500` instead.

**Note**: The current implementation does not handle some edge cases with streaming requests/responses. Those will require special handling. I will make a follow-up PR for that. See review comments for more details.

Risk Level:
Medium. 
- Technically a breaking change, but default envoy buffer limits are 1 MiB. Unlikely users will hit this limit with typical requests.
- Configurable by a runtime feature, `true` by default: `envoy.reloadable_features.grpc_json_transcoder_adhere_to_buffer_limits`

Testing:
- Unit tests for unary/streaming, request/response
- Integration tests for unary request/responses over buffer limit.
- Integration tests for streaming responses under/over buffer limit.

Docs Changes: None

Release Notes: Yes

Platform Specific Features: None

### CODE REVIEW

**Comment by nareddyt** [comment]
@qiwzhang Please take a quick look at the draft PR.
@htuch Can you give more context on the scary changes in b/155331763#comment11
cc @TAOXUY

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:885]: Not,  this is not what I read from this [flow control](https://github.com/envoyproxy/envoy/blob/main/source/docs/flow_control.md#decoder-filters)

**Review by qiwzhang — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:885]: Yes, this is the doc I referenced. I suggest starting with [this section](https://github.com/envoyproxy/envoy/blob/main/source/docs/flow_control.md#http1-and-http2-filters).

The main point is:
> If a buffer is a streaming buffer, i.e. the buffered data will resolve over time, it should return FilterDataStatus::StopIterationAndWatermark to pause further data processing, which will cause the ConnectionManagerImpl to trigger watermark callbacks on behalf of the filter. If a filter can not make forward progress without the complete body, it should return FilterDataStatus::StopIterationAndBuffer.

As mentioned in the PR description, the transcoder cannot make forward progress without the complete body:
> Flow control via watermarking will not help the filter: Data stuck in a buffer will never be transcoded until more data comes in. If watermarking occurs, then the data will be stuck forever.

Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:885]: Hence why I did not implement the watermarking. Perhaps someone else can confirm this?

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:537]: This is not very accurate.  incoming request bytes are converted on the flight to proto.  It is stored [here](https://github.com/grpc-ecosystem/grpc-httpjson-transcoding/blob/master/src/include/grpc_transcoding/request_message_translator.h#L128) if I read it correctly.

We may have to count its size.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (qiwzhang) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:885]: I see your points now.  So if we exceeded the buffer_limit, we have to bail out, waiting is not going to help.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:537]: Got it. Will spend some time evaluating if there are any other buffers internally.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (antoniovicente) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:885]: You may need to do something to generate an error reply when buffered data exceeds the allowed limit.  StopIterationNoBuffer won't produce the appropriate error response.

Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:885]: I has an offline discussion with @antoniovicente, should be resolved. FYI we call `decoder_callbacks_->sendLocalReply` just a few lines below to generate the error reply.

File: source/common/http/filter_manager.cc
Comment (nareddyt) [inline source/common/http/filter_manager.cc:192]: TODO: This ASSERT was introduced in https://github.com/envoyproxy/envoy/pull/7720. I need to check why. The assertion fails when we `sendLocalReply` in `encodeData` for a server streaming case.

Comment (nareddyt) [reply source/common/http/filter_manager.cc:192]: This change is part of #15075 now. Feel free to ignore the diff here.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:552]: I took a look at the transcoder library. This will only be a problem for the streaming cases. If two streaming messages occur in succession, then the byte limit will be inaccurate and the buffered data will exceed the allowed limit. This should be rare.

For unary request/response, the current implementation works for all cases. The buffer limit is checked before the message reaches any of those internal buffers. So the size unary messages will be counted correctly. 

I would like to move forward with this PR and follow-up later. The diff is getting large. I am not worried about introducing a partial implementation here. There is **never** a case where the transcoder will incorrectly reject a request. It will always count the correct buffer usage or (in the streaming case) undercount buffer usage.

**Review by qiwzhang — DISMISSED** [review]

**Comment by antoniovicente** [comment]
What does [DNS] refer to in the PR title?

**Comment by antoniovicente** [comment]
adding @alyssawilk who is more familiar with buffering inside filter pipeline and meaning of the various filter return statuses.

**Comment by nareddyt** [comment]
> What does [DNS] refer to in the PR title?

DO NOT SUBMIT. Renamed PR title to make it clearer.

As stated in the description:

> DO NOT SUBMIT until #15075 is merged
(ready for review)

I did not have a chance to take a closer look at review comments in #15075 yet. But this PR is final, so ready to review.

File: docs/root/version_history/current.rst
Comment (alyssawilk) [inline docs/root/version_history/current.rst:17]: what happened previously?  Do these changes need to be runtime guarded per CONTRIBUTING.md?

File: source/common/http/filter_manager.cc
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:186]: Ah, now I have the context for this change.
ccing @snowp for a second opinion but in mine, if we end up with a sendLocalReply in call stack of the filter chain, we shouldn't be calling commonHandleAfterDataCallback at all, we should short circuit and early return from FilterManager::decodeData

I'd be inclined to revert that PR and add that short circuit here, but again let's see what @snowp or @mattklein123 think.

**Review by alyssawilk — COMMENTED** [review]
Cool, thanks for tackling this issue.  Looks like you need a clang tidy check and a main merge, and I'll take another pass.  Pinging folks on that nasty filter chain issue in the interim though so hopefully we can get consensus on that one in parallel with sorting out the little things :-)

File: source/common/http/filter_manager.cc
Comment (mattklein123) [reply source/common/http/filter_manager.cc:186]: Sorry I don't think I have enough context. What PR are we reverting?

Comment (alyssawilk) [reply source/common/http/filter_manager.cc:186]: https://github.com/envoyproxy/envoy/pull/15075

Basically sendLocalReply is supposed to Just Work (TM) in all cases. If a filter calls sendLocalReply after the response filter chain has started, it should result in either the local reply being sent straight to the codec (as we can't traverse the response filter chain again safely) or stream reset (if the headers have been shipped downstream already)

Here in this PR we try to sendLocalReply under the stack of encodeData, and it turns it it doesn't work becasue the commonHandleAfterDataCallback assertions are failing.  #15075 was an attempt to fix the asserts to not break, but IMO we shouldn't call commonHandleAfterDataCallback if we've aborted the original reply (in encodedata) in favor of sendLocalReply being shipped immediately out.

Comment (mattklein123) [reply source/common/http/filter_manager.cc:186]: OK yeah +1 on reverting the other PR and then let's just sort it all out in this PR? I think that would be easier to review agreed.

File: docs/root/version_history/current.rst
Comment (nareddyt) [reply docs/root/version_history/current.rst:17]: This is a behavior change. Previously the filter would accept and transcode messages that went over the buffer limit.

This filter is **not** marked as alpha in the [thread model](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/threat_model), so it should qualify for runtime guarding.

I was under the impression that security fixes are not breaking changes, but correct me if I'm wrong. I can add a guard feature that defaults to `true`. Can any other @envoyproxy/maintainers comment on this?

Comment (alyssawilk) [reply docs/root/version_history/current.rst:17]: my guess is the change will then cause problems for folks sending large responses, so IMO it's worth runtime gurading and maybe even adding a pointer to config to change if it causes errors.  Happy if you want to hold off until someone else chimes in.

File: source/common/http/filter_manager.cc
Comment (alyssawilk) [reply source/common/http/filter_manager.cc:186]: I'd be inclined to track when end_stream=true is sent to the codec (probably in FilterManager::maybeEndEncode) and check against that?  Seems more general purpose.

Comment (nareddyt) [reply source/common/http/filter_manager.cc]: Added a new state variable to short-circuit the logic.

I tried finding a more general approach (by using `end_stream_`), but there are cases where end_stream=false but the data buffer still needs to be handled.

I do not have enough understanding of the FilterManager code to feel confident about making larger changes to it. If the current non-general approach works, can we go move forward with it?

File: docs/root/version_history/current.rst
Comment (nareddyt) [reply docs/root/version_history/current.rst]: Thanks, makes sense. Added the runtime guard and documented the breaking change in the version history. Added a couple of unit tests with the runtime feature disabled.

**Comment by nareddyt** [comment]
Ready for review

File: source/extensions/filters/http/grpc_json_transcoder/transcoder_input_stream_impl.cc
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/transcoder_input_stream_impl.cc:12]: unint64_t  or std::size_t?   is it ever be -1?

Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/transcoder_input_stream_impl.cc]: Done. `uint64_t` is correct.

File: source/common/http/filter_manager.cc
Comment (nareddyt) [reply source/common/http/filter_manager.cc]: Thanks! This makes much more sense and is generalized. Switched it to use `state_.destroyed_`.

Comment (nareddyt) [reply source/common/http/filter_manager.cc]: I had to move the short-circuit logic back into `commonHandleAfterDataCallback`. Short-circuiting before with the `state_.destroyed` broke a HTTP conn manager test for buffer limits, because it has to handle the `StopIterationAndBuffer` case.

https://github.com/envoyproxy/envoy/blob/e20957c011b11a4f9f935ec7556ed89d1d8c22d7/test/common/http/conn_manager_impl_test_2.cc#L1403

```

test/common/http/conn_manager_impl_test_2.cc:1433
Failed
Logs:
 [2021-03-11 18:49:07.433][14][trace][http] [source/common/http/filter_manager.cc:1190] [C71][S0] encode data called: filter=0xdb93f6ab810 status=1
 Do NOT contain:
  'debug', 'Resetting stream due to response_payload_too_large. Prior headers have already been sent'
Stack trace:
  0x1be5a18: Envoy::Http::HttpConnectionManagerImplTest_HitResponseBufferLimitsAfterHeaders_Test::TestBody()
  0x417abe4: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x416acbe: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x41530b3: testing::Test::Run()
  0x4153cad: testing::TestInfo::Run()
... Google Test internal frames ...

test/common/http/conn_manager_impl_test_2.cc:1428
Actual function call count doesn't match EXPECT_CALL(stream_, resetStream(_))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
Stack trace:
  0x4127fe9: testing::internal::GoogleTestFailureReporter::ReportFailure()
  0x1b521a8: testing::internal::Expect()
  0x412db6e: testing::internal::UntypedFunctionMockerBase::VerifyAndClearExpectationsLocked()
  0x22b3100: testing::internal::FunctionMocker<>::~FunctionMocker()
  0x22b0cd3: Envoy::Http::MockStream::~MockStream()
  0x1ca3d81: Envoy::Http::HttpConnectionManagerImplTest::~HttpConnectionManagerImplTest()
  0x1c53f88: Envoy::Http::HttpConnectionManagerImplTest_HitResponseBufferLimitsAfterHeaders_Test::~HttpConnectionManagerImplTest_HitResponseBufferLimitsAfterHeaders_Test()
  0x1c53fac: Envoy::Http::HttpConnectionManagerImplTest_HitResponseBufferLimitsAfterHeaders_Test::~HttpConnectionManagerImplTest_HitResponseBufferLimitsAfterHeaders_Test()
  0x416b69b: testing::Test::DeleteSelf_()
  0x417abe4: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x416acbe: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x4153cf8: testing::TestInfo::Run()
  0x415457c: testing::TestSuite::Run()
... Google Test internal frames ...
```

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (alyssawilk) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:543]: I'd be inclined to remove the checkIf in naming here so
if (decoderBufferLimitReached()) and (encoderBufferLimitReached())
as it reads a bit more smoothly.

**Review by alyssawilk — COMMENTED** [review]
this looks great!  just one final nit from me and you're good to go.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Done

**Review by alyssawilk — APPROVED** [review]
@snowp or @mattklein123 would one of you be up for final pass?

**Comment by mattklein123** [comment]
I can take a look.

File: docs/root/version_history/current.rst
Comment (mattklein123) [inline docs/root/version_history/current.rst:28]: Can you also document what buffer config a user would have to modify to fix this if they see rejected requests?

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (mattklein123) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: Why is the string in braces? Same below?

File: source/common/http/filter_manager.cc
Comment (mattklein123) [inline source/common/http/filter_manager.cc:190]: I would really prefer this change in a dedicated PR, but if it stays here please have a dedicated core test for this change.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, LGTM with small comments.

/wait

File: source/common/http/filter_manager.cc
Comment (nareddyt) [reply source/common/http/filter_manager.cc:190]: Re-opened #15075 as the dedicated PR. Integration test in that PR fails without this short-circuit logic.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: This follows the pre-existing format for Response Code Details. Other examples from across Envoy:
```
upstream_reset_before_response_started{remote_reset}
jwt_authn_access_denied{Jwt_is_missing}
grpc_json_transcode_failure{INVALID_ARGUMENT}
```

Comment (mattklein123) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: I'm not sure why the others are like that. I don't think the braces here add anything. Can they be removed?

/wait

File: docs/root/version_history/current.rst
Comment (nareddyt) [reply docs/root/version_history/current.rst:28]: Done.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: Hmm, I'd prefer not to change the existing format in this PR. This filter also has other `sendLocalReply` with the braces. Should we do this in a follow-up PR?

**Comment by nareddyt** [comment]
CI will fail until #15075 is merged.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (mattklein123) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: We can leave the existing ones but I don't see why we would add new ones like this.

/wait

Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: Ack, done.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (mattklein123) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]:
```suggestion
        absl::StrCat(RcDetails::get().GrpcTranscodeFailed, "request_buffer_size_limit_reached"));
```

Same below

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comment.

/wait

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: The current commit results in the following string: `grpc_json_transcode_failure_request_buffer_size_limit_reached`

The suggested change would result in: `grpc_json_transcode_failurerequest_buffer_size_limit_reached`

That doesn't seem right. Keep in mind this is in `absl::StrCat`, the entire string is passed to `details` arg of `sendLocalReply`.

https://github.com/envoyproxy/envoy/blob/cfdd59a6280fed47ef4b881491c9ed0e96d4476d/include/envoy/http/filter.h#L403

Comment (mattklein123) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: Ah OK I see, got it. Thanks.

Comment (mattklein123) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc:919]: I think I also understand now why it was using braces before, sorry. I'm fine either way. Are you good with this or do you want to go back to braces?

Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: No worries. I prefer the braces for consistency. Reverted the last commit.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
